rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+[prior-art]: #prior-art
+
+<!--
+Discuss prior art, both the good and the bad, in relation to this proposal. A few examples of what
+this can include are:
+
+- Can we learn something about this proposal from other projects in CTD?
+- Can we learn something about this proposal from other transit agencies?
+- Can we learn something about this proposal from past experiences in other jobs or projects?
+
+This section is intended to encourage you as an author to think about the lessons from other places,
+and provide readers of your RFC with a fuller picture.
+
+If there is no prior art, that is fine.
+
+Note that while precedent is some motivation, it does not on its own motivate an RFC.
+-->
+
+Workflow automation for Terraform is entirely new territory for the authors of this document. We've done a fair amount of research and factfinding to formulate a plan that we think brings CTD's usage of Terraform more in line with industry best practices, and takes advantage of products that take some of the work off our shoulders. That said, it's entirely possible there's something we missed, and we're open to recommendations of solutions, software or patterns from anyone who cares to weigh in.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+<!--
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature
+  before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the
+  future independently of the solution that comes out of this RFC?
+-->
+
+One major unresolved question is the timeline for implementing all the improvements laid out in this RFC. As noted in the [Drawbacks][drawbacks] section above, it's a lot of work. Once this RFC is approved we'll need to scope out the work and divide it accordingly between the infrastructure and engineering teams.
+
+Other questions for the reader:
+
+- Is the proposed [root module naming convention][root-module-naming-convention] sensible and sufficiently comprehensive?
+- Does the [isolated module testing workflow][reference-module-testing] make sense? Does it seem easy enough to carry out?
+- How aggressive should we be about [standardizing on infrastructure component modules][guide-standardization] for ECS and RDS across all applications/teams?
+- How aggressive should we be about [versioning and publishing application modules][guide-module-versioning] for open-source applications?
+- If changes to non-versioned application modules can spill over into other modules, as noted in the [Scalr Workflow for Module Testing][scalr-workflow-module-testing], is that enough of a reason to push to version and publish all application modules?
+- Does the decision to [delay splitting the `prod` module and instead version all production module calls with a git hash][scalr-workflow-production] make sense?
```

Comment:
Switching to using git hashes seems like a bit of a headache that has potential to go wrong, and it'd be nice to avoid thinking about versioning for application modules. Continuing without them is no different from the status quo, where we make and protect changes to application modules with the understanding that they affect prod.

If we do want to eventually move prod config into the application root module, then I don't think adding that to the scope adds too much risk to this RFC, and we should plan to do it rather than delay it indefinitely.

Reply:
> Switching to using git hashes seems like a bit of a headache that has potential to go wrong

I agree that it will add some overhead, but in what ways do you think it could go wrong?

> Continuing without them is no different from the status quo, where we make and protect changes to application modules with the understanding that they affect prod.

My concern is that the concurrent work to [add automation around Terraform](https://github.com/mbta/technology-docs/blob/idw-terraform-workflow-improvements/rfcs/0000-terraform-workflow-improvements.md#5-terraform-automation) will limit this understanding and increase the chance that module changes unintentionally impact production, which is a significant risk. I'm hoping we can strike a balance that requires less infrastructure team oversight on small changes while still ensuring that every change is deliberate and has a known impact. Leaving less to intuition is worth the additional step, IMO.

> If we do want to eventually move prod config into the application root module, then I don't think adding that to the scope adds too much risk to this RFC, and we should plan to do it rather than delay it indefinitely.

Fair point. I want to make sure we have the process and automation figured out before we take on production, but that doesn't have to mean calling all of production out of scope. I'll think more about this.